### PR TITLE
#788: list commits immediately on fs init

### LINKED ISSUE #788: Frequent errors when first starting Jupyter
https://dotscience.slack.com/archives/C7PHSRL82/p1578674668057000

@lukemarsden reports:

> I get this error fairly often when starting jupyter on a runner (when the runner didn’t yet have the project dot). always resolves when running again but it’s a bad look. is there some asynchrony in dotmesh knowing its commits after an initial pull/clone?

image

> in this case it’s the machine under my desk in Bristol, which is quite fast bare-metal machine in case that makes a difference

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #788 

Or not? I can't reproduce the problem, so I wrote a test and potential fix but I can't be sure it did anything. @lukemarsden can you see if this fixes it for you?

### CODE REVIEW

**Comment by itamarst** [comment]
Had a bug in previous version, so testing should await me getting tests passing.

**Comment by itamarst** [comment]
OK, fixed that I think.

File: cmd/dotmesh-server/rpc.go [hunk 1]
```diff
@@ -754,9 +754,16 @@ func (d *DotmeshRPC) Commits(
 	if err != nil {
 		return err
 	}
-	snapshots, err := d.state.SnapshotsForCurrentMaster(filesystemId)
-	if err != nil {
-		return err
+
+	var snapshots []Snapshot
+	for {
```
[the remark this answers is not in the document]
Comment (itamarst) [reply cmd/dotmesh-server/rpc.go]: Good point, yes, I'll fix that. Will still wait to merge until @lukemarsden actually tries this and sees if it his fixes his issue.

**Comment by lukemarsden** [comment]
FYI – InitFilesystemMachine is about initializing a state machine to check (and modify) the state of the system, not (necessarily) creating a filesystem.

https://github.com/dotmesh-io/dotmesh/issues/788#issuecomment-574758951

Filesystems can show up in a number of ways, they can be created explicitly on the host (there's a CreateFilesystem method around here https://github.com/dotmesh-io/dotmesh/blob/75f9c909d6da9ec67aa30d575024b0a6903cec73/cmd/dotmesh-server/controller.go#L613), they can be received in a new cross-cluster replication stream (push/pull/clone), they can be received in a cluster-local replication (replication, moving dots between nodes).

Dots (filesystems) with no commits (snapshots) on them are certainly valid system state, so I'm a bit concerned about the 10 second delay on Commits calls for that case. Do we create initial commits on new dots in dotmesh now? (I remember we did that from inside the gateway for a while, as for the dotscience use case we need an initial commit to be able to synchronize even an empty filesystem to jupyter etc).

The codepath I had in mind to try to fix this in was not the rpc side, where this feels like a band-aid (your diff takes a hint that filesystem state might not be up to date, and adds latency), but rather in the codepath that receives a cross-cluster replication stream. Can we ensure that we set up the snapshot list in memory before we return success from the (effective) `dm clone`?

**Comment by itamarst** [comment]
OK, going to try that approach in a different branch, starting with seeing if I can reproduce the bug with a test based off `dm clone`.

**Pull request closed without merging**